### PR TITLE
feat(connector-ubiquity): added new, initial implementation

### PR DESCRIPTION
**Pull request by RafaelAPB**
This plugin defines interfaces for developers to use a wrapped version of the Ubiquity SDK. Ubiquity is a high performance, multi-chain API for accessing blockchain data, i.e., provides one API to access multiple protocols. 
This API complements Cactus current connector offering by allowing to connect seamlessly to a multitude of public blockchains.

Signed-off-by: Rafael Belchior <rafael.belchior@tecnico.ulisboa.pt>

### CODE REVIEW

File: packages/cactus-plugin-ubiquity/src/main/typescript/prometheus-exporter/data-fetcher.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-ubiquity/src/main/typescript/prometheus-exporter/data-fetcher.ts:1]: The response.type.ts doesnt contain Transactions

File: packages/cactus-plugin-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts:84]: Is this piece of code expected to work in future PR's or added by mistake?

**Review by jagpreetsinghsasan — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-ubiquity/src/main/typescript/prometheus-exporter/data-fetcher.ts
Comment (RafaelAPB) [reply packages/cactus-plugin-ubiquity/src/main/typescript/prometheus-exporter/data-fetcher.ts]: Fixed!

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:152]: Will the URL's strictly start with `svc.blockdaemon.com` prefix?

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (jagpreetsinghsasan) [inline packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:154]: This commented code can be removed if not required

**Review by jagpreetsinghsasan — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (RafaelAPB) [reply packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:152]: Yes!

**Comment by RafaelAPB** [comment]
I made a PR to my branch because Plugin-Ubiquity requires an env. variable with an API secret: https://github.com/RafaelAPB/blockchain-integration-framework/actions/runs/3159805550

**Comment by jagpreetsinghsasan** [comment]
> I made a PR to my branch (the CI job passes) because Plugin-Ubiquity requires an env. variable with an API secret: [RafaelAPB/blockchain-integration-framework/actions/runs/3159805550](https://github.com/RafaelAPB/blockchain-integration-framework/actions/runs/3159805550)

Can we not limit the CI till the point the need for that env var arises? (Or maybe move this requirement in code till the stage its required). Apart from that, LGTM

**Review by jagpreetsinghsasan — APPROVED** [review]

**Comment by takeutak** [comment]
@jagpreetsinghsasan Now the reviewer is only you, but could you need other's review?  (The merge of pull-requests needs at least 2 maintainers' approval)

**Comment by jagpreetsinghsasan** [comment]
> @jagpreetsinghsasan Now the reviewer seems to be only you, but could you need another's review? (The merge of pull-requests needs at least 2 maintainers' approval)

Oh, I am not sure why others reviewers aren't added, but its good to have reviews from other maintainers as others might know some aspects of the project and the requirements better.
Adding other maintainers for review.

**Comment by jagpreetsinghsasan** [comment]
> @jagpreetsinghsasan what do you mean by limit the CI? I've synced with @petermetz and unfortunately there's no secure way to share secrets such that the plugin can run successfully on Cactus CI.

If there is no other way, we can go with the current approach and maybe raise this as a feature with ubiquity for its inclusion (Not sure if that's needed as I am unaware of ubiquity, just a suggestion)

**Review by takeutak — APPROVED** [review]

**Comment by takeutak** [comment]
@RafaelAPB  Thanks for contributing!  Very sorry for my late approval.

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:128]: @RafaelAPB Please use camelCase for method names where possible.

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:111]: @RafaelAPB Please use camelCase for method names where possible.

File: packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/src/main/typescript/ubiquity.ts:89]: @RafaelAPB Please use camelCase for method names where possible.

File: packages/cactus-plugin-ubiquity/src/main/typescript/generated/openapi/typescript-axios/.gitignore
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/src/main/typescript/generated/openapi/typescript-axios/.gitignore:4]: @RafaelAPB Please delete this file - it is ignored in the open api generator ignore file as well.

File: packages/cactus-plugin-ubiquity/src/main/typescript/generated/openapi/typescript-axios/.npmignore
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/src/main/typescript/generated/openapi/typescript-axios/.npmignore:1]: @RafaelAPB Please delete this file - it is ignored in the open api generator ignore file as well.

**Review by petermetz — CHANGES_REQUESTED** [review]
@RafaelAPB Please delete the package-lock file, it's redundant with the yarn.lock file in the project root directory.

File: packages/cactus-plugin-ubiquity/package.json
Comment (petermetz) [inline packages/cactus-plugin-ubiquity/package.json:3]: @RafaelAPB Please bump this to 1.1.0 and also bump the dependencies versions that are cactus packages (because we have issued a release in the meantime)

**Review by petermetz — CHANGES_REQUESTED** [review]

**Comment by RafaelAPB** [comment]
done @petermetz

**Comment by petermetz** [comment]
Suggested edit:
```diff
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/endpoints/transaction/get-transactions-endpoint.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/endpoints/transaction/get-transactions-endpoint.ts
index 26c29df2..bd532039 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/endpoints/transaction/get-transactions-endpoint.ts
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/endpoints/transaction/get-transactions-endpoint.ts
@@ -15,13 +15,13 @@ import {
 
 import { registerWebServiceEndpoint } from "@hyperledger/cactus-core";
 
-import { PluginUbiquity } from "../../ubiquity";
+import { PluginLedgerConnectorUbiquity } from "../../plugin-ledger-connector-ubiquity";
 
 import OAS from "../../../json/openapi.json";
 
 export interface IGetTransactionsByAddressEndpointOptions {
   logLevel?: LogLevelDesc;
-  ubiquity: PluginUbiquity;
+  ubiquity: PluginLedgerConnectorUbiquity;
 }
 
 export class GetTransactionsByAddressEndpoint implements IWebServiceEndpoint {
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/ubiquity.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/plugin-ledger-connector-ubiquity.ts
similarity index 82%
rename from packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/ubiquity.ts
rename to packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/plugin-ledger-connector-ubiquity.ts
index 80393e6b..76a3a163 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/ubiquity.ts
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/plugin-ledger-connector-ubiquity.ts
@@ -1,6 +1,4 @@
 import type { Express } from "express";
-import type { Server as SecureServer } from "https";
-import type { Server } from "http";
 import {
   Logger,
   Checks,
@@ -9,13 +7,12 @@ import {
 } from "@hyperledger/cactus-common";
 import { PrometheusExporter } from "./prometheus-exporter/prometheus-exporter";
 import { v4 as uuidV4 } from "uuid";
-import { AxiosResponse } from "axios";
+import type { AxiosResponse } from "axios";
 import fetch from "node-fetch";
 import dotenv from "dotenv";
 import path from "path";
 
-import * as nodeFetch from "node-fetch";
-import { Optional } from "typescript-optional";
+import type { Response } from "node-fetch";
 import OAS from "../json/openapi.json";
 
 import { NETWORK_TYPE, PROTOCOL_TYPE } from "./helpers/api-parameters";
@@ -36,7 +33,7 @@ import {
 
 const envPath = path.join(__dirname, "../.env");
 dotenv.config({ path: envPath });
-export interface IPluginUbiquity {
+export interface IPluginLedgerConnectorUbiquityOptions {
   logLevel?: LogLevelDesc;
   authToken: string;
   instanceId?: string;
@@ -44,27 +41,17 @@ export interface IPluginUbiquity {
   prometheusExporter?: PrometheusExporter;
 }
 
-// Plugin connector interface not implemented at this moment, so required interfaces are empty
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-interface DeployContractV1Request {}
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-interface DeployContractV1Response {}
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-interface RunTransactionRequest {}
-// eslint-disable-next-line @typescript-eslint/no-empty-interface
-interface RunTransactionResponse {}
-
-export class PluginUbiquity
+/**
+ * Note: The plugin connector interface methods are not provided by the initial
+ * implementation, and because of that the generic type parameters are set to
+ * `never`
+ */
+export class PluginLedgerConnectorUbiquity
   implements
-    IPluginLedgerConnector<
-      DeployContractV1Request,
-      DeployContractV1Response,
-      RunTransactionRequest,
-      RunTransactionResponse
-    >,
+    IPluginLedgerConnector<never, never, never, never>,
     ICactusPlugin,
     IPluginWebService {
-  public static readonly CLASS_NAME = "PluginUbiquity";
+  public static readonly CLASS_NAME = "PluginLedgerConnectorUbiquity";
   private readonly instanceId: string;
   private readonly log: Logger;
   private client: UbiquityClient;
@@ -73,10 +60,10 @@ export class PluginUbiquity
   private endpoints: IWebServiceEndpoint[] | undefined;
 
   public get className(): string {
-    return PluginUbiquity.CLASS_NAME;
+    return PluginLedgerConnectorUbiquity.CLASS_NAME;
   }
 
-  constructor(public readonly opts: IPluginUbiquity) {
+  constructor(public readonly opts: IPluginLedgerConnectorUbiquityOptions) {
     const fnTag = `${this.className}#constructor()`;
     expect(process.env.UBIQUITY_AUTH_TOKEN).toBeTruthy();
     const authToken = process.env.UBIQUITY_AUTH_TOKEN;
@@ -105,10 +92,10 @@ export class PluginUbiquity
 
     this.prometheusExporter.startMetricsCollection();
   }
-  deployContract(options?: string): Promise<DeployContractV1Response> {
+  deployContract(options?: string): Promise<never> {
     throw new Error(`Method not implemented. ${options}`);
   }
-  transact(options?: string): Promise<RunTransactionResponse> {
+  transact(options?: string): Promise<never> {
     throw new Error(`Method not implemented. ${options}`);
   }
   getConsensusAlgorithmFamily(): Promise<ConsensusAlgorithmFamily> {
@@ -180,7 +167,7 @@ export class PluginUbiquity
   public async nativeGetTx(
     protocol: PROTOCOL_TYPE,
     network: NETWORK_TYPE,
-  ): Promise<nodeFetch.Response> {
+  ): Promise<Response> {
     const fnTag = `${this.className}:NativeGetTx`;
     this.log.debug("enter ", fnTag);
     const ethereumUrl = `https://svc.blockdaemon.com/${protocol}/${network}/native`;
@@ -250,10 +237,6 @@ export class PluginUbiquity
     return this.endpoints;
   }
 
-  public getHttpServer(): Optional<Server | SecureServer> {
-    return Optional.empty();
-  }
-
   public getOpenApiSpec(): unknown {
     return OAS;
   }
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/public-api.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/public-api.ts
index c0dc9271..db90bf84 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/public-api.ts
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/public-api.ts
@@ -1,2 +1,6 @@
-export { PluginUbiquity, IPluginUbiquity } from "./ubiquity";
+export {
+  PluginLedgerConnectorUbiquity,
+  IPluginLedgerConnectorUbiquityOptions,
+} from "./plugin-ledger-connector-ubiquity";
+
 export * from "./generated/openapi/typescript-axios/index";
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/init.test.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/init.test.ts
index 1dbf0e61..d5da82d2 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/init.test.ts
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/init.test.ts
@@ -12,8 +12,8 @@ import {
 import { LogLevelDesc, Checks } from "@hyperledger/cactus-common";
 
 import {
-  PluginUbiquity,
-  IPluginUbiquity,
+  PluginLedgerConnectorUbiquity,
+  IPluginLedgerConnectorUbiquityOptions,
 } from "../../../main/typescript/public-api";
 
 const testCase = "initialize ubiquity plugin";
@@ -40,12 +40,12 @@ describe(testCase, () => {
   });
 
   test(testCase, async () => {
-    const options: IPluginUbiquity = {
+    const options: IPluginLedgerConnectorUbiquityOptions = {
       logLevel: logLevel,
       authToken: authToken,
       instanceId: uuidv4(),
     };
-    const api = new PluginUbiquity(options);
+    const api = new PluginLedgerConnectorUbiquity(options);
     Checks.truthy(api);
   });
 });
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts
index 30293d12..a5528f4b 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts
@@ -11,8 +11,8 @@ import {
 import { LogLevelDesc, Checks } from "@hyperledger/cactus-common";
 
 import {
-  PluginUbiquity,
-  IPluginUbiquity,
+  PluginLedgerConnectorUbiquity,
+  IPluginLedgerConnectorUbiquityOptions,
 } from "../../../main/typescript/public-api";
 
 const testCase = "obtain all txs";
@@ -40,14 +40,14 @@ describe(testCase, () => {
     await expect(pruning).resolves.toBeTruthy();
   });
 
-  const options: IPluginUbiquity = {
+  const options: IPluginLedgerConnectorUbiquityOptions = {
     logLevel: logLevel,
     authToken: authToken,
     instanceId: uuidv4(),
     basePath: "https://svc.blockdaemon.com/universal/v1/",
   };
 
-  const api = new PluginUbiquity(options);
+  const api = new PluginLedgerConnectorUbiquity(options);
   Checks.truthy(api);
 
   // Rainbow bridge ethereum custodian contract

```

**Comment by petermetz** [comment]
> done @petermetz

@RafaelAPB Most of the renaming was missing from the plugin file. I sent a patch.

File: packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/ubiquity.ts
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-ubiquity/src/main/typescript/ubiquity.ts:38]: @RafaelAPB Executing code outside of class methods is highly discouraged because it can cause crashes at import-time that are very hard to debug when it happens on the CI or in production.
Please move this code to either the constructor or the `onPluginInit` async handler depending on what is the best place for it (I think the constructor is fine because the documentation of the methods called do not say that it would be async)

**Review by petermetz — CHANGES_REQUESTED** [review]

**Comment by RafaelAPB** [comment]
@petermetz I incorporated the changes. Should be good for merging. If possible before the workshop, would be amazing!

**Review by petermetz — CHANGES_REQUESTED** [review]
> @petermetz I incorporated the changes. Should be good for merging. If possible before the workshop, would be amazing!

@RafaelAPB Sorry for the slow response. Some of the changes from the patch are missing, are you sure you applied it all? (It can be applied automatically as long as you haven't changed anything in the meantime that would make it conflict)

**Comment by RafaelAPB** [comment]
> > @petermetz I incorporated the changes. Should be good for merging. If possible before the workshop, would be amazing!
> 
> @RafaelAPB Sorry for the slow response. Some of the changes from the patch are missing, are you sure you applied it all? (It can be applied automatically as long as you haven't changed anything in the meantime that would make it conflict)

@petermetz Which changes are missing? I changed all ``cactus-plugin-ubiquity`` to ``cactus-plugin-ledger-connector-ubiquity``

**Comment by petermetz** [comment]
> > > @petermetz I incorporated the changes. Should be good for merging. If possible before the workshop, would be amazing!
> > 
> > 
> > @RafaelAPB Sorry for the slow response. Some of the changes from the patch are missing, are you sure you applied it all? (It can be applied automatically as long as you haven't changed anything in the meantime that would make it conflict)
> 
> @petermetz Which changes are missing? I changed all `cactus-plugin-ubiquity` to `cactus-plugin-ledger-connector-ubiquity`

@RafaelAPB Attaching a screenshot from my patch above (see comments above). Some of these changes are missing for example. If you apply that patch as-is it should take care of it.

image

**Review by petermetz — CHANGES_REQUESTED** [review]
^^

**Comment by RafaelAPB** [comment]
@petermetz sorry for the miss! I sent a patch, hopefully it would be good to go now!

**Comment by petermetz** [comment]
Suggested edit:
```diff
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts b/packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtain-all-txs.test.ts
similarity index 100%
rename from packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtainAllTxs.test.ts
rename to packages/cactus-plugin-ledger-connector-ubiquity/src/test/typescript/integration/obtain-all-txs.test.ts

```

**Comment by RafaelAPB** [comment]
@petermez patch applied!

**Comment by RafaelAPB** [comment]
@petermetz patch applied!

**Comment by RafaelAPB** [comment]
@petermetz done!

**Comment by petermetz** [comment]
Suggested edit:
```diff
diff --git a/packages/cactus-plugin-ledger-connector-ubiquity/package.json b/packages/cactus-plugin-ledger-connector-ubiquity/package.json
index fa37a6aa..d5d045f2 100644
--- a/packages/cactus-plugin-ledger-connector-ubiquity/package.json
+++ b/packages/cactus-plugin-ledger-connector-ubiquity/package.json
@@ -1,6 +1,6 @@
 {
     "name": "@hyperledger/cactus-plugin-ledger-connector-ubiquity",
-    "version": "1.1.0",
+    "version": "1.1.2",
     "description": "Allows Cactus nodes to connect to a set of public blockchains.",
     "keywords": [
         "Hyperledger",

```

**Review by petermetz — CHANGES_REQUESTED** [review]
@RafaelAPB Last two things:

1. bump the version of the package to the current latest release version (I sent a suggestion just now)
2. Make sure the commit message also contains the information that you put in the github PR description.

After these, we should be good to go (unless v1.1.3 gets released in the meantime in which case I'll have to bug you again to bump that here as well before we merge - I don't think that'll be the case but just letting you know in case it does happen)

**Comment by RafaelAPB** [comment]
@petermetz good to go!

**Review by petermetz — APPROVED** [review]
@RafaelAPB The PR title was not updated still. I changed it just now. We are good to go from my side, just need to figure out what's wrong with tests.

**Comment by RafaelAPB** [comment]
> cactus-plugin-connector-ubiquity

@petermetz  The tests from cactus-plugin-connector-ubiquity fail because the API is not setup. They pass on my codespace: https://github.com/RafaelAPB/blockchain-integration-framework/actions/runs/3573530344/jobs/6007894998

The Iroha tests also pass on my space, so I would assume it is something regarding the resources of the CI server? Perhaps we can trigger another run of these tests?

**Comment by petermetz** [comment]
> @petermetz just to confirm - is this ready to be merged?

@RafaelAPB We just need to do some due diligence on the test. Can you confirm that the token you are using on your fork is the same that you gave me to set up for the upstream (this) repo? If yes, then the best course of action probably is to just merge it and hope that the token will become accessible once the code is on the main branch and work our way back from there if this turns out to be false.

**Comment by petermetz** [comment]
Oh wait, nvm, I see in the logs that the token is not set at all. Let's just merge this and hope that once it's on the main branch the code will have access to the token in the GH repo secret. And if not, then we can just iterate further on it knowing that something else is wrong.

**Pull request merged**
